api/comments/serializers.py

```diff
@@ -0,0 +1,155 @@
+    ])
+
+    id = IDField(source='_id', read_only=True)
+    type = TypeField()
+    content = ser.CharField()
+
+    user = JSONAPIHyperlinkedRelatedField(view_name='users:user-detail', lookup_field='pk', lookup_url_kwarg='user_id', link_type='related', read_only=True)
+    node = JSONAPIHyperlinkedRelatedField(view_name='nodes:node-detail', lookup_field='pk', lookup_url_kwarg='node_id', link_type='related', read_only=True)
+    target = JSONAPIHyperlinkedGuidRelatedField(link_type='related', meta={'type': 'get_target_type'})
+    replies = JSONAPIHyperlinkedIdentityField(view_name='comments:comment-replies', lookup_field='pk', link_type='self', lookup_url_kwarg='comment_id')
+    reports = JSONAPIHyperlinkedIdentityField(view_name='comments:comment-reports', lookup_field='pk', lookup_url_kwarg='comment_id', link_type='related', read_only=True)
+
+    date_created = ser.DateTimeField(read_only=True)
+    date_modified = ser.DateTimeField(read_only=True)
+    modified = ser.BooleanField(read_only=True, default=False)
+    deleted = ser.BooleanField(read_only=True, source='is_deleted', default=False)
+
+    # LinksField.to_representation adds link to "self"
+    links = LinksField({})
+
+    class Meta:
+        type_ = 'comments'
+
+    def create(self, validated_data):
+        user = self.context['request'].user
+        node_id = self.context['view'].kwargs.get('node_id', None)
+        target_id = self.context['view'].kwargs.get('comment_id', None)
+
+        if node_id:
+            node = self.context['view'].get_node()
+            target = node
+        elif target_id:
+            target = Comment.load(target_id)
+            node = target.node
+
+        validated_data['user'] = user
+        validated_data['node'] = node
+        validated_data['target'] = target
+        now = datetime.utcnow()
+        validated_data['date_created'] = now
+        validated_data['date_modified'] = now
+
+        if node.can_comment(Auth(user)):
+            comment = Comment(**validated_data)
+            comment.save()
+        else:
+            raise PermissionDenied("Not authorized to comment on this project.")
+
+        return comment
+
+    def update(self, comment, validated_data):
+        assert isinstance(comment, Comment), 'comment must be a Comment'
+        auth = Auth(self.context['request'].user)
+        if validated_data:
+            if validated_data.get('content', None):
+                comment.edit(validated_data['content'], auth=auth, save=True)
+            is_deleted = validated_data.get('is_deleted', None)
+            if is_deleted:
+                comment.delete(auth, save=True)
+            elif comment.is_deleted:
+                comment.undelete(auth, save=True)
+        return comment
+
+    def get_target_type(self, obj):
+        target_id = obj._id
+        target = Guid.load(target_id).referent
+        if isinstance(target, Node):
+            return 'node'
+        elif isinstance(target, Comment):
```

Comment:
ValidationErrors are nice for when user input values for a particular field fail to match a specific pattern, but we don't generally raise them when the problem is with something with the entire resource requested as a whole. For the two cases here:

1. If no node of that ID is found, perhaps a 404 (not found)?
2. If the node-or-comment-only view gets something that is neither a node or a comment, something has gone horribly wrong, and a 400 ("omg-why-are-you-doing-that") seems appropriate.